alacritty/src/renderer/solidrect.rs

```diff
@@ -0,0 +1,200 @@
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut vbo);
+            gl::GenBuffers(1, &mut ebo);
+
+            gl::BindVertexArray(vao);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, ebo);
+
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+            // Position.
+            gl::VertexAttribPointer(
+                0,
+                2,
+                gl::FLOAT,
+                gl::FALSE,
+                (size_of::<Vertex>()) as _,
+                ptr::null(),
+            );
+            gl::EnableVertexAttribArray(0);
+
+            // Color
+            gl::VertexAttribPointer(
+                1,
+                4,
+                gl::UNSIGNED_BYTE,
+                gl::TRUE,
+                (size_of::<Vertex>()) as _,
+                offset_of!(Vertex, color) as *const _,
+            );
+            gl::EnableVertexAttribArray(1);
+
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Self { vao, vbo, ebo, indices: Vec::new(), vertices: Vec::new(), uploaded_indices: 0 }
+    }
+
+    pub fn draw(
+        &mut self,
+        program: &RectShaderProgram,
+        size_info: &SizeInfo,
+        rects: Vec<RenderRect>,
+    ) {
+        if rects.is_empty() {
+            return;
+        }
+
+        // Prepare common state
+        unsafe {
+            // Remove padding from viewport.
+            gl::Viewport(0, 0, size_info.width() as i32, size_info.height() as i32);
+
+            gl::Enable(gl::BLEND);
+            gl::BlendFuncSeparate(gl::SRC_ALPHA, gl::ONE_MINUS_SRC_ALPHA, gl::SRC_ALPHA, gl::ONE);
+
+            // Setup bindings. VAO will set up attribs and EBO, but not VBO.
+            gl::BindVertexArray(self.vao);
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        for rect in &rects {
+            self.append_rect(center_x, center_y, rect);
+        }
+
+        self.draw_accumulated();
+    }
+
+    fn append_rect(&mut self, center_x: f32, center_y: f32, rect: &RenderRect) {
+        if rect.alpha <= 0. {
+            return;
+        }
+
+        assert!(self.vertices.len() <= MAX_U16_INDICES - 4);
+
+        // Calculate rectangle position.
+        let x = (rect.x - center_x) / center_x;
+        let y = -(rect.y - center_y) / center_y;
+        let width = rect.width / center_x;
+        let height = rect.height / center_y;
+        let color = Rgba {
+            r: rect.color.r,
+            g: rect.color.g,
+            b: rect.color.b,
+            a: (rect.alpha * 255.) as u8,
+        };
+
+        self.vertices.extend_from_slice(&[
+            Vertex { x, y, color },
+            Vertex { x, y: y - height, color },
+            Vertex { x: x + width, y, color },
+            Vertex { x: x + width, y: y - height, color },
+        ]);
```

Comment:
This is a bit of a strange way to push four values to a vector. I'd hope that the optimizer cleans this up for you, but it's still a bit weird.

I think a better idea would be to just push four times. I'd assume you might have done it this way to prevent it from reallocating more than once? Could maybe just `self.vertices.reserve(4);` be good enough to solve that problem?

Reply:
Honestly, I don't know the best way to do it. I'll try to investigate it separately when I have time.

My mental model here is `std::vector` from C++. For vector it is more optimal than pushing items one by one, and definitely faster than reserving each time. Reserving messes with growth strategy: it makes sure that there's an allocation **each time** an element is added, while default growth in push_back will allocate only occasionally by pre-allocating more storage.

Reply:
There's no `emplace_back` in rust, however I'm not really sure that push is better than what you're doing when it comes to copy members. One can benchmark it, but that's about it. Though, push is more common in such cases afaics.

Reply:
I left it as is for now. Can change to individual pushes if you insist. Or if I find time to benchmark it somehow and have data to back either approach.

Reply:
To be clear, my suggestion was not focused on performance but on code clarity.

I'm relatively certain that `extend_from_slice` is decent for performance especially with small datatypes. However it does invoke `clone` which **requires** you to be aware of these datatypes and make sure they actually stay small and easy to clone around.

The assumption here is that you **do not** have ownership over the slice, which is why it's not a mutable borrow. So the code doesn't really express our intentions very well, because cloning an existing slice certainly isn't our goal here, but our goal is to push new elements, that were just created, onto that vector.

So unless there's a good reason to do it this way (as in: significant performance benefit), our code should focus on being clear on what it's trying to do and help with maintenance in the future. Currently this code doesn't do that very well.

Reply:
I see what you mean. Will change to pushes.